### PR TITLE
Created IA for German phone prefixes

### PR DESCRIPTION
**Pull request by oliverrahner**
https://duck.co/ia/view/german_phone_prefixes

### CODE REVIEW

**Comment by daxtheduck** [comment]
@oliverrahner Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

While the team and the community review your submission, we'd like to make sure you look good! If you'd like a free DuckDuckGo tshirt, please send over the following info to merch@duckduckgo.com, and be sure to put "Dev Merch" in the subject line:

Name:
Street Address:
City:
State / Province:
Country:
Phone (Non-US Orders only): 
T-shirt size:
GitHub username:
Link to this Pull Request:

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: t/GermanPhonePrefix.t [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+zci answer_type => "german_phone_prefix";
+zci is_cached   => 1;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::GermanPhonePrefix )],
+    # At a minimum, be sure to include tests for all:
+    # - primary_example_queries
+    # - secondary_example_queries
+    'vorwahl 05491'         => test_zci(
```
Comment (GuiltyDolphin) [inline t/GermanPhonePrefix.t]: @oliverrahner Could we perhaps clear up the formatting a bit?

Comment (oliverrahner) [inline t/GermanPhonePrefix.t]: Missed those, fixed and updated.

**Comment by moollaza** [comment]
@oliverrahner Thanks for contributing and welcome to DuckDuckHack!

Looking at this IA, we're expecting very specific inputs, and have very simple outputs to display, based on a static data set (I take it this data isn't going to change much or at all?)

This is actually an ideal use-case for a Fathead Instant Answer. Would you be interested in converting this?

You could borrow most of what you're written here, except it would need to take in your input file and produce a tab-delimited file. We'd then use this to create a DB where the keys are the expected queries.

Docs are available here: http://docs.duckduckhack.com/resources/fathead-overview.html

An example IA can be seen here: https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/share/fathead/celeb_heights/parse.pl

What do you think?

**Comment by oliverrahner** [comment]
@moollaza 
I tried to grasp the concept of fathead answers.
If I understood correctly, the only two things of "real" importance are
- output.txt and
- FatHeadName.pm

Right?

Because, as you guessed correctly, the data is unlikely to change at all over the course of several years (the last version is from the end of 2009).
Also, the responsible German agency (the Bundesnetzagentur; http://www.bundesnetzagentur.de/cln_1931/DE/Sachgebiete/Telekommunikation/Unternehmen_Institutionen/Nummerierung/Rufnummern/ONVerzeichnisse/ONB_Verzeichnis/ONBVerz_Basepage.html) did release a csv file containing all relevant data but for some reason, this official data source is missing all letter accents ("München" becomes "Munchen", whereby the correct transcription would be "Muenchen").
So, the main work that I was still planning to do is to improve the datasource quality which is mainly a manual process.

Long story short, I could imagine that a valid approach would be to deliver output.txt and GermanPhonePrefix.pm as a Fathead IA without the need for any fetching or parsing scripts.

**Comment by moollaza** [comment]
> Long story short, I could imagine that a valid approach would be to deliver output.txt and GermanPhonePrefix.pm as a Fathead IA without the need for any fetching or parsing scripts.

@oliverrahner sorry for the delay! The reason we like to have scripts available to parse the data is so we can ideally run the script at a later date to update it. Seeing as you said this isn't likely to change though. I *think* providing just an output.txt will be ok.

@jdorweiler do you know if the other FatHead files are required?

**Comment by jdorweiler** [comment]
@oliverrahner You can just include the output.txt file.  The .pm file isn't needed anymore.

**Comment by moollaza** [comment]
@oliverrahner Ping! I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

There's absolutely no rush, just wanted to make sure you're not stuck on anything.

**Comment by moollaza** [comment]
@oliverrahner I'm going to close this for now. If you're still interested in finishing it up, please let me know!

**Comment by edgesince84** [comment]
@oliverrahner,  thanks again for creating the phone prefixes IA! You’ve done a really good job here so far and we'd like you to help make DuckDuckGo even better for yourself and others interested in German phone number locations.

When you're ready to move on, please: 

- Create a new Instant Answer for searches you do related to Germany

OR

- Test live <topic> Instant Answers (or those in development) for bugs and suggest improvements

Detailed instructions explaining how you can help are available at the top of this [page](https://github.com/duckduckgo/duckduckgo/wiki/Creating-or-Improving-Instant-Answers).

Thanks again!
